New REMOVE_RANGE command for range removal

saurtron opened the pull request:
### Work done

- New REMOVE_RANGE command, removing ranges by cmdId or Tag.

### Work yet to be done in this PR

- Sort out callin behaviour, if any
  - kind of decided, but can still think about it.
- Maybe refactor to better share code with the cmdid/tag loop
  - not sure it's going to be pretty due to inverse looping and 'active' inside the cmdid/tag loop.

### Related issues

- https://github.com/beyond-all-reason/spring/issues/1082
- https://github.com/beyond-all-reason/Beyond-All-Reason/issues/2190

Also see following "precursor" prs:
- https://github.com/beyond-all-reason/spring/pull/2137
- https://github.com/beyond-all-reason/spring/pull/2154

### Remarks

- It's a bit unclear how callins should react 
  - how is gui supposed to be notified? I don't see the other loop doing any callins, but gui needs them to be notified of changes, specially for factories, otherwise the gui queue numbers don't get updated without some special gui code.
  - for now decided on actually calling UnitCmdDone for the remove command.

### REMOVE_RANGE specs

- ALT: remove by index or tag switch
- CONTROL: alternative queue selection.
  - for factories alternative queue is the factory command queue, default queue is the newUnitCommands queue.
  - for other units no effect.

callins: After successful execution, it will call UnitCmdDone.

Examples:
  - `spGiveOrderToUnit(unitID, CMD.REMOVE, nil, {'meta', 'ctrl'}) -- delete everything`
  - `spGiveOrderToUnit(unitID, CMD.REMOVE, 2, {'meta', 'ctrl', 'alt'}) -- delete all but the 1st element`
  - `spGiveOrderToUnit(unitID, CMD.REMOVE, {1, 20}, {'meta', 'ctrl', 'alt'}) -- delete from 1st to 2th`
  - `spGiveOrderToUnit(unitID, CMD.REMOVE, {startTag, endTag}, {'meta', 'ctrl'}) -- delete from position of startTag, to endTag`

saurtron committed:
Add META modifier to INSERT cmd. With it will do range deletion, instead of by cmdid/tag.

sprunk reviewed (commented):
Looking nice.

saurtron committed:
Check firstIndex isn't negative.

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
Don't META if ALT is present too for now.

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
Run UnitCmdDone callin for finished REMOVE commands.

sprunk reviewed (commented):
Looks great.

saurtron committed:
Implement meta+alt mode (range where start and end provided as tags).

saurtron commented:
Ended up implementing tag mode too, it's a bit verbose and who knows if someone needs it, but actually when not providing params it doesn't matter if ALT is used since it defaults to whole queue anyways, otherwise yes, ALT is indexes, no ALT is tags.

Let me know if you think it's too much can always remove, but felt its incomplete otherwise, besides, we were not considering the case where no ALT, but no params too. In that case it works just like with ALT.

saurtron committed:
Make sure lastIndex is greater or equal to firstIndex.

saurtron committed:
Move foundStart, foundEnd a bit up.

saurtron commented:
Another consideration, indexes here are c++ based, so first is 0, last is queue.size()-1. Should we maybe use lua indexes?

sprunk commented:
> Should we maybe use lua-like indexes instead?

Good point, should use Lua 1-based.

> Ended up implementing tag mode too

Great.

saurtron committed:
No need to reset iterator here.

saurtron committed:
Factor out FindTagIndex from GetRemoveLimitsFromOptions.

saurtron committed:
Use 1 based start and end indexes.

saurtron committed:
Use implicit optional notation and -- instead of -=.

saurtron committed:
No need to use 'active' here.

saurtron committed:
some spacing.

saurtron committed:
refresh iterator since it may be invalidated.

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1212,6 +1212,49 @@ void CCommandAI::ExecuteInsert(const Command& c, bool fromSynced)
 	SlowUpdate();
 }
 
+const std::optional<int> CCommandAI::FindTagIndex(const CCommandQueue& queue, unsigned int tag) const
```
`const` in return type probably doesn't make a difference, maybe this was supposed to be `std::optional <const int>`?

saurtron replied:
hmm, not sure in this case, maybe even any const is superfluous since if not present I guess whoever calls it sets the constness or not depending on how they will store it.

if const, tho, regarding const inside or outside, not sure if correct, but reading here https://stackoverflow.com/questions/55060223/is-the-content-of-a-const-stdoptional-always-const

"const std::optional<T> and const std::optional<const T> are effectively the same"

sprunk replied:
If it works as-is then it's probably fine either way.

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1212,6 +1212,49 @@ void CCommandAI::ExecuteInsert(const Command& c, bool fromSynced)
+			firstIndex = std::max<int>(c.GetParam(0)-1, firstIndex);
+		if (c.GetNumParams() >= 2)
+			lastIndex = std::min<int>(c.GetParam(1)-1, lastIndex);
+	} else if (c.GetNumParams() > 0) {
```
Slightly inconsistent where all other checks in this function use `>= (N+1)`  but here it's `> N`, consider
```suggestion
	} else if (c.GetNumParams() >= 1) {
```

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1212,6 +1212,49 @@ void CCommandAI::ExecuteInsert(const Command& c, bool fromSynced)
+		if (c.GetNumParams() >= 2)
+			lastIndex = std::min<int>(c.GetParam(1)-1, lastIndex);
+	} else if (c.GetNumParams() > 0) {
+		auto firstTagIndex = FindTagIndex(queue, c.GetParam(0));
```
```suggestion
		const auto firstTagIndex = FindTagIndex(queue, c.GetParam(0));
```

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1212,6 +1212,49 @@ void CCommandAI::ExecuteInsert(const Command& c, bool fromSynced)
+			return std::nullopt;
+
+		if (c.GetNumParams() >= 2) {
+			auto lastTagIndex = FindTagIndex(queue, c.GetParam(1));
```
```suggestion
			const auto lastTagIndex = FindTagIndex(queue, c.GetParam(1));
```

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1238,11 +1281,51 @@ void CCommandAI::ExecuteRemove(const Command& c)
+		int nElements = lastIndex - firstIndex + 1;
+
+		CCommandQueue::iterator ci = queue->begin() + lastIndex;
+		while(nElements > 0) {
```
```suggestion
		while (nElements > 0) {
```

sprunk commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1238,11 +1281,51 @@ void CCommandAI::ExecuteRemove(const Command& c)
 	}
 
-	if ((c.GetNumParams() <= 0) || (queue->size() <= 0))
+	if ((c.GetNumParams() < 0) || (queue->size() <= 0))
```
The <0 check looks worrisome, can NumParams actually be negative and what would that mean?

saurtron replied:
don't think so tbh, it's unsigned after all and code also has nowhere it could be negative even if it wasn't unsigned, was wondering the same thing, also queue->size() don't think that can be < 0.

no idea why it's like that.

sprunk replied:
Sounds safe to just
```suggestion
	if (queue->empty())
```

sprunk reviewed (commented):
Looks good, perhaps it's a good moment to undraft.

saurtron committed:
No need for utility header.

saurtron committed:
use >= for consistency

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
make firstTagIndex const

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
make lastTagIndex const

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
spacing after while

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron committed:
just check if queue is empty, checking GetNumParams() < 0 doesn't make sense.

Co-authored-by: sprunk <spr.ng@o2.pl>

sprunk commented:
>  users are known to create queues with 10s of thousands of elements.
> Anyways, we might need to somehow limit max queue size, since otherwise it's a way to DOS the game.

```lua
function gadget:AllowCommand(unitID, _, _, cmdID)
  return Spring.GetUnitCommandCount(unitID) < 5 -- brood war allows 5 so should be enough for everyone
      or isWhitelisted[cmdID] -- put stuff like cmd.remove, state commands etc there
end
```

saurtron commented:
> ```lua
> function gadget:AllowCommand(unitID, _, _, cmdID)
>   return Spring.GetUnitCommandCount(unitID) < 5 -- brood war allows 5 so should be enough for everyone
> ```

True, although to limit to five you'd have to refine a bit more since some commands can create at least 100 other commands in the queue :D, also it could be a different queue depending on the unit, command and modifiers :D.

saurtron committed:
Index parsing fixes.

- Don't accept indexed start bigger than queue size, or end smaller than start.
- Only swap start and end for tag based indexing.

saurtron commented:
A couple extra things I noticed and would like to discuss:

1) Best approach to validating and clamping input parameters

Noticed some issues in how we were "validating" the input parameters.

at the moment we're clamping and also inverting start and end if end index is smaller.

I think some cases shouldn't be clamped, like if start index is smaller than queue size, it's unlikely the user wants to clamp that to queue size, since, say you want to "delete everything but the first command", passing {2} as parameter, if that would be clamped on a queue of 1 it would result in clearing the queue, thus deleting out of range.

So, I think we should clamp for convenience, but only in a way where the final range is smaller than the initial range, that means we can clamp end index to queue size. Start index shouldn't really be clamped, unless negative.

Regarding negative indexes, not convinced we should clamp those, maybe better to just fail the command.

Same goes with inverting start and end, I see a small case for this when referencing through tags, but when referencing through indexes there's no reason they should be provided inverted. For now I left inverting only for tags, but even in that case not really sure.

2) Behaviour when clearing the whole queue

I noticed a "glitch/bug", when clearing the whole queue with this, in some cases (depending on cmdID) the first command doesn't get completely cancelled, for example if building something, the builder will actually cancel the command, but will continue targeting the buildee, and automatically apply guard and repair later on (not sure where this comes from, but not from IdleUnit callins). This seems to be the case also for previous behavior of deleting by cmdID or tag filters (so it doesn't seem to depend on order of looping). It doesn't happen when leaving some extra commands, since then the builder will actually proceed to those instead of applying further logic, likely why this was never noticed with REMOVE.

Found two ways to fix this, although not sure it's wanted, or just let it be and maybe fixing later, feel like that leaves a bug here tho, although a "bug" that already existed in this command.

1- Factor whole queue code from "[add command without shift](https://github.com/beyond-all-reason/spring/blob/3003a30f988ce7c13c0060f108a69efd383996da/rts/Sim/Units/CommandAI/CommandAI.cpp#L999)" (at least when `queue == commandQue`). And call that when detecting clearing of the whole queue. Easy to do when deleting through indexes as in this PR work, and few lines extra since it refactors existing code. "Impossible" to do when deleting through cmdid/tag. Could always call it just for the last element in that case, or leave that unresolved.
2- Instead of deleting the last command, issue a GiveAllowedCommand(Command(CMD_STOP)), that will clear it, but basically the same as calling the code from (1) but just for the last element :P, and with the side effect of pushing yet another STOP command into the queue at the end (that will then immediately be processed by SlowUpdate called at the end).

Imo (1) is a nice way of fixing this, with the additional side effect of greatly optimizing whole queue clearing, although that's not really a use case for this command. Planning to refactor that code later also to allow clearing whole ranges, I think it's possible and could be a great optimization in general (and maybe fix the need to the CMD_STOP hax).

saurtron committed:
Also run CmdDone when returning early.

Co-authored-by: sprunk <spr.ng@o2.pl>

saurtron commented on rts/Sim/Units/CommandAI/CommandAI.cpp:
```diff
@@ -1238,11 +1284,51 @@ void CCommandAI::ExecuteRemove(const Command& c)
+	if (c.GetOpts() & META_KEY) {
+		const auto limits = GetRemoveLimitsFromOptions(c, *queue);
+		if (!limits)
+			return;
```
Yeah, unsure what to do there... could maybe run it only in that case, or either maybe run the callin in all return early cases but that'd be a bit cumbersome. Even if considered "valid", since it doesn't do anything, do you really think it needs the UnitCmdDone callin there? Not sure it's really needed if nothing really happened, unless we want to send it unconditionally so the sender can always do smth after this is resolved one way or another...

sprunk replied:
I would prefer if it produced an event because it feels arbitrary that a valid command would just silently disappear with no trace if it just so happened its conditions were already fulfilled. Perhaps check what happens if you give a move command onto the unit's current position, or a "set firestate to holdfire" command if it's already holdfire etc. (I don't know what happens but I would assume they both produce an event), and then make remove behave consistently with that.

saurtron replied:
Tricky since state commands never send any CmdDone, and move probably always sends since it wouldn't consider same position as a fail condition.

Maybe ATTACK is a good example, it does send CmdDone through FinishCommand when returning early because of any bad parameters.

On the other hand, INSERT and REMOVE didn't generally send any callins, there could be some cases where queue manipulation led to other commands ending and sending CmdDone themselves.

Anyways I'm leaning more towards adding CmdDone in all early returns too, just to be generally consistent inside CMD_REMOVE, like attack does. I think not having something here was probably a problem/bug, and regarding INSERT.

Will have to check how lua finds out about the queue changing, but there doesn't seem to be a good mechanism for the general case. I guess it finds out because it knows the command is about to be given, but not because the command was already processed.

sprunk commented:
> I think some cases shouldn't be clamped, like if start index is smaller than queue size, it's unlikely the user wants to clamp that to queue size, since, say you want to "delete everything but the first command", passing {2} as parameter, if that would be clamped on a queue of 1 it would result in clearing the queue, thus deleting out of range.

Yes. If the param is {2} then commands 0 and 1 should stay intact, always. AFAICT that's how it currently works so it's good.

> So, I think we should clamp for convenience, but only in a way where the final range is smaller than the initial range, that means we can clamp end index to queue size. Start index shouldn't really be clamped, unless negative. I already changed to work like that

Yes, great.

> Regarding negative indexes, not convinced we should clamp those, maybe better to just "fail" the command.

Perhaps starting index `-n` should mean `n`th from the end (like it does e.g. for `Spring.GetUnitCurrentCommand`). So you could do {-3} to mean "remove last 3" without knowing the size of the queue too, otherwise you'd have to know the size and do {size-3}. I'm not sure either way though.

> Same goes with inverting start and end, I see a small case for this when referencing through tags, but when referencing through indexes there's no reason they should be provided inverted. For now I left inverting only for tags, but even in that case not really sure.

Sounds safest not to invert and consider that an invalid command imo.

> Behaviour when clearing the whole queue

IMO if the weird behaviour already happens with the existing modes of REMOVE then keep it as-is for now, and fix it in a separate PR. Some function to "poke" units to flush any cached orders they think they might have and refresh their CommandAI would be good to have, then it could be called directly after REMOVE to fix this and also could be exposed to Lua to fix #126.

saurtron committed:
Never accept lastIndex smaller than firstIndex, or firstIndex greater than queue size.

saurtron commented:
> Perhaps starting index `-n` should mean `n`th from the end (like it does e.g. for `Spring.GetUnitCurrentCommand`). So you could do {-3} to mean "remove last 3" without knowing the size of the queue too, otherwise you'd have to know the size and do {size-3}. I'm not sure either way though.

Yes, also thought about this, although not sure if implementing now, or leave it for later.

Also, if both are negative, first should also be smaller, right, or should we invert the meaning, ie: `{-3, -6}` or `{-6, -3}` xd, also the clamping behaviour should invert...

> Sounds safest not to invert and consider that an invalid command imo.

Ok, removed the swapping and made that illegal in all situations.

> IMO if the weird behaviour already happens with the existing modes of REMOVE then keep it as-is for now, and fix it in a separate PR. Some function to "poke" units to flush any cached orders they think they might have and refresh their CommandAI would be good to have, then it could be called directly after REMOVE to fix this and also could be exposed to Lua to fix #126.

Then we will have to recommend CMD_STOP for full queue flushing (but not for factories clearing the factory queue, since they can't do it).

The problem with this is I didn't find a good way to perform that flush, otherwise I'd add it here XD, anyways yes, it can be fixed later on and for now just leave the weird behaviour be.

saurtron committed:
Improve comment.

saurtron committed:
Move removeByID down, since it's not used in the top loop.

It's name also applies only to the second loop.

saurtron committed:
Remove unused braces.

saurtron committed:
Also move active down, since it's not used in the first loop.

saurtron committed:
Fix top of loop comments and move active.

sprunk commented:
> I didn't find a good way to perform that flush after the fact, otherwise I'd add it here XD

I'm fairly sure that giving two CMD_WAIT commands works as a flush. Perhaps check whether it helps in this case and if yes, what exactly happens inside that makes it work.

badosu commented:
Is there a way we can document these behaviors? I'm afraid it is prone to be bunched together with all other obscure modifier behaviors on CMDs that currently are mostly unknown or used in siloed niches.

Reading the description I only could figure out what this was supposed to do after finding the examples, even after a brief look at the code (no comments).

_(note: is this up to 2nd or up to 20th? unsure if theres a typo or not `spGiveOrderToUnit(unitID, CMD.REMOVE, {1, 20}, {'meta', 'ctrl', 'alt'}) -- delete from 1st to 2th`)_

At least the entry to changelog should help

saurtron commented:
> Ok, I just realized there's #2172 addressing my documentation concern, I removed my previous comments on this post.
> 
> _(note: is this up to 2nd or up to 20th? unsure if theres a typo or not `spGiveOrderToUnit(unitID, CMD.REMOVE, {1, 20}, {'meta', 'ctrl', 'alt'}) -- delete from 1st to 2th`)_

Oh yeah, right that's why I wrote 2th originally, but then corrected to 2nd... good catch. Fixed now.

saurtron commented:
As noted [here](https://github.com/beyond-all-reason/spring/pull/2172#discussion_r2039721802), this might be making the REMOVE cmd api a bit too overloaded, and it could be better to create a separate command instead, like CMD_REMOVE_RANGE for example.

sprunk commented:
Splitting sounds good. I wouldn't do `REMOVE_RANGE` because there's multiple kinds of range (tags, indexes) and would instead stick to making them accept one type of value. So adding `CMD_REMOVE_BY_INDEX`, `CMD_REMOVE_BY_TAG`, and `CMD_REMOVE_BY_CMDID` sounds good. Remove-multiple (i.e. range) could be a modifier (either alt or meta, idk) for INDEX/TAG, and perhaps CMDID could also remove one by default and remove all with the remove-multiple modifier.

saurtron commented:
> Splitting sounds good. I wouldn't do `CMD_REMOVE_RANGE` because there's multiple kinds of range (tags, indexes) and would instead stick to making them accept one type of value. So adding `CMD_REMOVE_BY_INDEX`, `CMD_REMOVE_BY_TAG`, and `CMD_REMOVE_BY_CMDID` sounds good. Remove-multiple (i.e. range) could be a modifier (either alt or meta, idk) for INDEX/TAG, and perhaps CMDID could also remove one by default and remove all with the remove-multiple modifier.

Personally, don't see that path, since then we'd have 3 commands with overlapping uses, ie, `CMD_REMOVE`, `CMD_REMOVE_BY_TAG` and `CMD_REMOVE_BY_CMDID`. People would still be using the old `CMD_REMOVE` since why would they bother to change it. Then we'd still need the modifier in all commands, to select range or filtering.

I much prefer `CMD_REMOVE` and `CMD_REMOVE_RANGE`, then we keep `CMD_REMOVE` as is, and just add `CMD_REMOVE_RANGE` doing the range removal, together with the same modifier the other command uses for selecting tag/index. We end with two commands to do everything and nothing is really lost.

The other way we end with 4 commands each with different modifiers and behaviour.

Option A)

- CMD_REMOVE -> uses alt to use tag or id
- CMD_REMOVE_BY_TAG -> uses alt to range or filter
- CMD_REMOVE_BY_INDEX -> uses alt to range or filter
- CMD_REMOVE_BY_CMDID -> default to filter (no range option since doesn't make much sense imo)

Option B)

- CMD_REMOVE -> uses alt to use tag or id
- CMD_REMOVE_RANGE -> maybe no options, since by index is only useful option, or else use alt to use tag or index

Imo (B) much better and we don't pollute the command list so much. Also CMD names much nicer imo.

Also keep in mind we added the tag option for range just because REMOVE already supported that and otherwise would be incomplete, but don't think that's going to have much use, so if creating CMD_RANGE we could just not add tag support, and wait to see if it's smth needed later on.

sprunk commented:
> Also keep in mind we added the tag option for range just because REMOVE already supported that and otherwise would be incomplete, but don't think that's going to have much use

Tag is important for polishing because it avoids latency-based race condition when given from unsynced.

> People would still be using the old CMD_REMOVE since why would they bother to change it.

Yes, but that is fine. We can deprecate it and encourage people to use the dedicated ones. There's got to be like 20 uses of cmd.remove total across the whole ecosystem.

> together with the same modifier the other command uses for selecting tag/index. We end with two commands to do everything (...) we don't pollute the command list so much

As far as I understand, having complex commands that do everything and change parameter meaning based on modifiers is why @rhys-vdw suggested the split-off in the first place. If we're using an opportunity to have primitive commands then we shouldn't do the bare minimum, in fact I maybe we could even have 7+ commands (existing remove, by tag, by tag range, by index, by index range, by cmdid all, by cmdid single) where none of them take any modifier (other than ctrl for queue choice).

saurtron commented:
Well sorry, but I don't see it at all, it's as bad to have too many options as too many commands (pollutes the command space, documentation and code too), I'll do the CMD_REMOVE_RANGE and we can see.

saurtron committed:
Move the range deletion remove to new command CMD_REMOVE_RANGE.

saurtron commented:
Implemented CMD_REMOVE_RANGE and changed PR description and title accordingly.

sprunk commented:
The remaining minor changes for regular remove could be in a separate PR since they seem uncontroversial.

saurtron commented:
> The remaining minor changes for regular remove could be in a separate PR since they seem uncontroversial.

Actually they are controversial XD. If left here would require a bit of a refactor. That's because in some places the cpp code just does uses the REMOVE command to remove a command with a specific tag from the queue.

Wanted to say that but you beat me to it.

saurtron committed:
Remove UnitCmdDone calling from REMOVE.

saurtron committed:
Restore previous ExecuteRemove code.

saurtron committed:
Fix comment about ALT behaviour on ExecuteRemoveRange.

rhys-vdw commented:
> Well sorry, but I don't see it at all, it's as bad to have too many options as too many commands (pollutes the command space, documentation and code too), I'll do the CMD_REMOVE_RANGE and we can see.

I disagree with this because the options don't map to keystrokes in this case. I see no reason to use the options argument which has already been flagged as kind of outdated cruft. I actually think it makes more sense to move away from `options` in most, if not all cases.

Even when the command does related to keystrokes e.g. when adding unit orders to a factory queue, it makes more sense to check for keybindings and send a different quantity e.g. send `5` rather than send unit ID + ctrl. The use of this pattern has resulted in difficulties remapping inputs such a shift to queue vs shift to place a line of buildings.

I think @sprunk and I are on the same page here:

> If we're using an opportunity to have primitive commands then we shouldn't do the bare minimum, in fact I maybe we could even have 7+ commands (existing remove, by tag, by tag range, by index, by index range, by cmdid all, by cmdid single) where none of them take any modifier that modifies the meaning of parameters (only ctrl to choose which queue the removal applies to).

Even the queue selection could be passed as a param, but any movement away from using options is a good choice IMO.

saurtron committed:
Remove tag option from REMOVE_RANGE.

saurtron commented:
Ok, removed the tag option, since that seems to be controversial, also lua method to GetIndexFromTag or so could make it uneeded, so maybe better to create that one afterwards.

Don't really see creating lots of commands with ugly names, and also don't see creating lots of similarly named methods, I think that also makes the api ugly and confusing to use, not just having options.

Regarding options themselves, above I described a mechanism to make them look on-point from lua, maybe some day we can do that or smth in that direction, personally doesn't bother me too much, since there's more pressing issues tbh.

I don't want to set a new command standard here, just implement a simple Range command. Currently commands have simple names and sometimes several options, that's how it works now pretty much. Trying to come up with a new standard in every PR is a bad idea imo, and when one is implemented, everything should be changed swiftly, otherwise we'd just have several standards coexist and later before one is implemented I'm sure the goals will move again and there'll be a new standard and so on.

I'd be willing to maybe make this a lua method, instead of (or in addition to) a command, that would save some network bytes, but wouldn't be usable from widgets. Other than that I think the feature is `Remove(index range)`.

saurtron commented:
> So the new API is just `REMOVE_RANGE` with `"ctrl"` changing whether it's the main or alt queue?

Yes, I'll add a PR afterwards with GetCommandTagIndex or smth like that, so we can have the tag version "for free".



> I'm still confused what the index of a tag would be. Is it `FindFirstTaggedCommandInQueue`? Or are tags unique? What if you wanted to find the last tag?

Was noted elsewhere, but the tag for a command is a unique id identifying it in it's queue. So when you know the tag of a command you can query or address it directly through that.

saurtron commented:
@sprunk what about merging this one?

sprunk commented:
I wanted to write my own alternative and let somebody else choose between the two.
